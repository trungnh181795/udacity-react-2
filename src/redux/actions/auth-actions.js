import { LOGOUT_AUTHED_USER, SET_AUTHED_USER } from "../constants/auth-constants";

export const setAuthedUser = (authedUser) => {
    return {
        type: SET_AUTHED_USER,
        authedUser,
    };
}

export const logoutAuthedUser = () => {
    return {
        type: LOGOUT_AUTHED_USER,
    };
}

export const handleOnLogin = (username, password) => (dispatch, getState) => {
    const { users } = getState();

    const user = Object.values(users).find((user) => user.id === username && user.password === password);

    if (!!user) {
        return dispatch(setAuthedUser(user));
    }
}

export const handleOnLogout = () => {
    return (dispatch) => {
        return dispatch(logoutAuthedUser());
    };
}