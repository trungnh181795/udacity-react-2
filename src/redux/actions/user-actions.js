import { ADD_ANSWER_USER, ADD_QUESTION_USER, RECEIVE_USERS } from "../constants/user-constants";

export const receiveUsers = (users) => {
    return {
        type: RECEIVE_USERS,
        users,
    };
}

export const addAnswerUser = (authedUser, qid, answer) => {
    return {
        type: ADD_ANSWER_USER,
        authedUser,
        qid,
        answer,
    };
}

export const addQuestionUser = ({ author, id }) => {
    return {
        type: ADD_QUESTION_USER,
        author,
        qid: id,
    };
}