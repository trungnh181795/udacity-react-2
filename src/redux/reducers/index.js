import { combineReducers } from "redux";
import authReducer from './auth-reducer';
import userReducer from './user-reducer';
import questionsReducer from './question-reducer';

export default combineReducers({
    authedUser: authReducer,
    users: userReducer,
    questions: questionsReducer,
});