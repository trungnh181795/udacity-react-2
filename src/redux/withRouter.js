import { useParams, /* Other hooks */ } from "react-router-dom";

const withRouter = Component => props => {
  const params = useParams();
  // Other hooks, useLocation, useNavigate, etc..
  return <Component {...props} {...{ params, /* other injected props */ }} />;
};

export default withRouter;