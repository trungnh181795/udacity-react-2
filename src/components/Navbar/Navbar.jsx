import { Link } from "react-router-dom";
import { connect } from "react-redux";
import { handleOnLogout } from "../../redux/actions/auth-actions";
import { AppBar, Box, Button, Container, Divider, Drawer, IconButton, ListItemIcon, ListItemText, MenuItem, MenuList, Paper, Stack, Toolbar, Typography } from "@mui/material";
import MenuIcon from "@mui/icons-material/Menu"
import AdbIcon from "@mui/icons-material/Adb";
import { useState } from "react";
import ArrowBackIosIcon from '@mui/icons-material/ArrowBackIos';
import { ContentCopy, ContentCut, ContentPaste } from "@mui/icons-material";

const Navbar = ({ dispatch, authedUserId }) => {
    const [open, setOpen] = useState(false)

    const logout = (e) => {
        e.preventDefault();
        dispatch(handleOnLogout());
    };

    return (
        <>
            <AppBar position="static" sx={{ height: '70px', display: 'flex', alignItems: 'center', justifyContent: 'center' }}>
                <Container maxWidth="xl">
                    <Toolbar disableGutters sx={{ display: 'flex', justifyContent: 'space-between' }}>
                        <IconButton
                            onClick={() => setOpen(true)}
                            size="large"
                            edge="start"
                            color="#FFFFFF"
                            aria-label="menu"
                            sx={{ mr: 2, display: { xs: 'block', lg: 'none' } }}
                        >
                            <MenuIcon color="#FFFFFF" />
                        </IconButton>
                        <Box sx={{ display: { xs: "none", md: "flex" }, alignItems: 'center' }}>
                            <AdbIcon sx={{ mr: 1 }} />
                            <Typography
                                variant="h5"
                                noWrap
                                component={Link}
                                to="/"
                                sx={{
                                    mr: 2,
                                    fontWeight: 700,
                                    color: "#FFFFFF",
                                    textDecoration: "none",
                                }}
                            >
                                Employee Polls
                            </Typography>
                        </Box>
                        <Box sx={{ display: { xs: 'none', lg: 'flex' }, alignItems: 'center' }}>
                            <Typography sx={{ color: '#FFFFFF', padding: '4px 8px', fontWeight: 700 }} component={Link} to="/">
                                Home
                            </Typography>
                            <Typography sx={{ color: '#FFFFFF', padding: '4px 8px', fontWeight: 700 }} component={Link} to="/leaderboard">
                                Leaderboard
                            </Typography>
                            <Typography sx={{ color: '#FFFFFF', padding: '4px 8px', fontWeight: 700 }} component={Link} to="/new">
                                New Poll
                            </Typography>
                        </Box>
                        <Stack direction="row" alignItems='center'>
                            <Typography
                                variant="h6"
                                component="span"
                                data-testid="user-information" sx={{ color: "#FFFFFF" }}>User: {authedUserId}</Typography>
                            <Button variant="contained" sx={{ bgcolor: '#FFFFFF', color: 'primary.main', ml: 1 }} onClick={logout}
                                className="font-medium px-3 py-2 text-slate-700 rounded-lg hover:bg-slate-100 hover:text-slate-900">Logout
                            </Button>
                        </Stack>
                    </Toolbar>
                </Container>
            </AppBar>
            <Drawer sx={{ width: '300px' }} anchor="left" open={open} onClose={() => setOpen(false)}>
                <Paper sx={{ width: 320, maxWidth: '100%', idth: '300px', height: '100%', bgcolor: '#23262F', borderRadius: 0 }}>
                    <Stack direction="row" justifyContent="flex-end" alignItems="center" sx={{ width: '100%', padding: '4px' }}>
                        <IconButton onClick={() => setOpen(false)}
                            size="large"
                            edge="start"
                            color="#FFFFFF"
                            aria-label="menu"
                            sx={{ mr: 2, display: { xs: 'block', lg: 'none' } }}>
                            <ArrowBackIosIcon color="#FFFFFF" />
                        </IconButton>
                    </Stack>
                    <Divider />
                    <MenuList>
                        <MenuItem onClick={() => setOpen(false)} component={Link} sx={{ color: '#FFFFFF' }} to="/">
                            <ListItemIcon>
                                <ContentCut fontSize="small" />
                            </ListItemIcon>
                            <ListItemText>Home</ListItemText>
                        </MenuItem>
                        <MenuItem onClick={() => setOpen(false)} component={Link} sx={{ color: '#FFFFFF' }} to="/leaderboard">
                            <ListItemIcon>
                                <ContentCopy fontSize="small" />
                            </ListItemIcon>
                            <ListItemText>Leaderboard</ListItemText>
                        </MenuItem>
                        <MenuItem onClick={() => setOpen(false)} component={Link} sx={{ color: '#FFFFFF' }} to="/new">
                            <ListItemIcon>
                                <ContentPaste fontSize="small" />
                            </ListItemIcon>
                            <ListItemText>New Poll</ListItemText>
                        </MenuItem>
                    </MenuList>
                </Paper>
            </Drawer>
        </>
    );
};

const mapStateToProps = ({ authedUser }) => ({
    authedUserId: authedUser.id,
});


export default connect(mapStateToProps)(Navbar);