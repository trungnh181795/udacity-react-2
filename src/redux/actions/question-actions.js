import { saveQuestion, saveQuestionAnswer } from "../../api/api";
import { ADD_ANSWER_QUESTION, ADD_QUESTION, RECEIVE_QUESTIONS } from "../constants/questions-constants";
import { addAnswerUser, addQuestionUser } from "./user-actions";

export const receiveQuestions = (questions) => {
    return {
        type: RECEIVE_QUESTIONS,
        questions,
    };
}

const addQuestion = (question) => {
    return {
        type: ADD_QUESTION,
        question,
    };
}

const addAnswerQuestion = (author, qid, answer) => {
    return {
        type: ADD_ANSWER_QUESTION,
        author,
        qid,
        answer,
    };
}

export const handleAddQuestion = (firstOption, secondOption) => {
    return async (dispatch, getState) => {
        const { authedUser } = getState();

        return saveQuestion(firstOption, secondOption, authedUser)
            .then((question) => {
                dispatch(addQuestion(question));
                dispatch(addQuestionUser(question))
            })
    };
}

export const handleAddAnswer = (questionId, answer) => {
    return async (dispatch, getState) => {
        const { authedUser } = getState();
        return saveQuestionAnswer(authedUser.id, questionId, answer)
            .then(() => {
                dispatch(addAnswerQuestion(authedUser.id, questionId, answer));
                dispatch(addAnswerUser(authedUser.id, questionId, answer));
            });
    };
}