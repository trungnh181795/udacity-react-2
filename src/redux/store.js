import { legacy_createStore } from "redux";
import reducers from "./reducers";
import middleware from "./middleware";

export const store = legacy_createStore(reducers, middleware)