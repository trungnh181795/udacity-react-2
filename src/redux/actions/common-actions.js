import { getInitialData } from "../../api/api";
import { receiveQuestions } from "./question-actions";
import { receiveUsers } from "./user-actions";

export const handleInitialData = () => {
    return async (dispatch) => {
        return getInitialData().then(({users, questions}) => {
            dispatch(receiveUsers(users));
            dispatch(receiveQuestions(questions));
        });
    };
}