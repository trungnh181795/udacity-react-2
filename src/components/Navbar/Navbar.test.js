/* eslint-disable testing-library/prefer-screen-queries */
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import React from "react";
import { setAuthedUser } from "../../redux/actions/auth-actions";
import { store } from "../../redux/store";
import Navbar from './Navbar';

describe("Navbar", () => {
    it("should render the view", () => {
        store.dispatch(setAuthedUser({ id: "sarahedo", password: "" }));
        const view = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Navbar />
                </BrowserRouter>
            </Provider>
        );
        expect(view).toBeDefined();
        expect(view).toMatchSnapshot();
    });

    it("should display username of logged-in user", () => {
        store.dispatch(setAuthedUser({ id: "sarahedo", password: "password123" }));

        const view = render(
            <Provider store={store}>
                <BrowserRouter>
                    <Navbar />
                </BrowserRouter>
            </Provider>
        );

        const userSpanElement = view.getByTestId("user-information");
        expect(userSpanElement.textContent).toBe("User: sarahedo");

    });
});