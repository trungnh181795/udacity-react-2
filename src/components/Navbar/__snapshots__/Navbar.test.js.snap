// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should render the view 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-1pxt6wy-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
        >
          <div
            class="MuiToolbar-root MuiToolbar-regular css-1gapj9v-MuiToolbar-root"
          >
            <button
              aria-label="menu"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-color#FFFFFF MuiIconButton-edgeStart MuiIconButton-sizeLarge css-1nocz5q-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-color#FFFFFF MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="MenuIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <div
              class="MuiBox-root css-13420go"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-elhp89-MuiSvgIcon-root"
                data-testid="AdbIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M5 16c0 3.87 3.13 7 7 7s7-3.13 7-7v-4H5v4zM16.12 4.37l2.1-2.1-.82-.83-2.3 2.31C14.16 3.28 13.12 3 12 3s-2.16.28-3.09.75L6.6 1.44l-.82.83 2.1 2.1C6.14 5.64 5 7.68 5 10v1h14v-1c0-2.32-1.14-4.36-2.88-5.63zM9 9c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1zm6 0c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1z"
                />
              </svg>
              <a
                class="MuiTypography-root MuiTypography-h5 MuiTypography-noWrap css-fmt68n-MuiTypography-root"
                href="/"
              >
                Employee Polls
              </a>
            </div>
            <div
              class="MuiBox-root css-oz7qqx"
            >
              <a
                class="MuiTypography-root MuiTypography-body1 css-zjrend-MuiTypography-root"
                href="/"
              >
                Home
              </a>
              <a
                class="MuiTypography-root MuiTypography-body1 css-zjrend-MuiTypography-root"
                href="/leaderboard"
              >
                Leaderboard
              </a>
              <a
                class="MuiTypography-root MuiTypography-body1 css-zjrend-MuiTypography-root"
                href="/new"
              >
                New Poll
              </a>
            </div>
            <div
              class="MuiStack-root css-1d9cypr-MuiStack-root"
            >
              <span
                class="MuiTypography-root MuiTypography-h6 css-jvdvyi-MuiTypography-root"
                data-testid="user-information"
              >
                User: 
                sarahedo
              </span>
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium font-medium px-3 py-2 text-slate-700 rounded-lg hover:bg-slate-100 hover:text-slate-900 css-ehj26b-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                Logout
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-1pxt6wy-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthXl css-19r6kue-MuiContainer-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular css-1gapj9v-MuiToolbar-root"
        >
          <button
            aria-label="menu"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-color#FFFFFF MuiIconButton-edgeStart MuiIconButton-sizeLarge css-1nocz5q-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-color#FFFFFF MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="MenuIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <div
            class="MuiBox-root css-13420go"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-elhp89-MuiSvgIcon-root"
              data-testid="AdbIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M5 16c0 3.87 3.13 7 7 7s7-3.13 7-7v-4H5v4zM16.12 4.37l2.1-2.1-.82-.83-2.3 2.31C14.16 3.28 13.12 3 12 3s-2.16.28-3.09.75L6.6 1.44l-.82.83 2.1 2.1C6.14 5.64 5 7.68 5 10v1h14v-1c0-2.32-1.14-4.36-2.88-5.63zM9 9c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1zm6 0c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1z"
              />
            </svg>
            <a
              class="MuiTypography-root MuiTypography-h5 MuiTypography-noWrap css-fmt68n-MuiTypography-root"
              href="/"
            >
              Employee Polls
            </a>
          </div>
          <div
            class="MuiBox-root css-oz7qqx"
          >
            <a
              class="MuiTypography-root MuiTypography-body1 css-zjrend-MuiTypography-root"
              href="/"
            >
              Home
            </a>
            <a
              class="MuiTypography-root MuiTypography-body1 css-zjrend-MuiTypography-root"
              href="/leaderboard"
            >
              Leaderboard
            </a>
            <a
              class="MuiTypography-root MuiTypography-body1 css-zjrend-MuiTypography-root"
              href="/new"
            >
              New Poll
            </a>
          </div>
          <div
            class="MuiStack-root css-1d9cypr-MuiStack-root"
          >
            <span
              class="MuiTypography-root MuiTypography-h6 css-jvdvyi-MuiTypography-root"
              data-testid="user-information"
            >
              User: 
              sarahedo
            </span>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium font-medium px-3 py-2 text-slate-700 rounded-lg hover:bg-slate-100 hover:text-slate-900 css-ehj26b-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Logout
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
